state-chain/traits/src/lib.rs

```diff
@@ -1223,6 +1223,28 @@ pub trait MinimumDeposit {
 	fn get(asset: Asset) -> AssetAmount;
 }
 
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, Clone, Copy, PartialEq, Eq, RuntimeDebug)]
+pub enum CallInfoId<AccountId> {
+	Pool { account: AccountId, base_asset: Asset },
+}
+
+// Creates an identifier that is used to group calls that should be scaled together.
+// Returns None if the call should not be scaled.
+pub trait FeeScalingCallInfoIdentifier<Call, AccountId, Balance> {
+	fn call_info_and_spam_prevention_upfront_fee(
+		_call: &Call,
+		_caller: &AccountId,
+	) -> Option<(CallInfoId<AccountId>, Balance)> {
+		None
+	}
+}
```

Comment:
I'm actually not sure this needs to care about the balance and account id at all. The fee is constant, it doesn't depend on the call. The account is simply passed through. 

All we need is function that decides which calls need to be scaled. Even the identifier can be totally opaque. 

```suggestion
pub trait CallIndexer<Call> {
	type CallIndex: Encode;
	fn call_index(
		_call: &Call,
	) -> Option<CallIndex> {
		None
	}
	
	fn opaque_call_index(
		call: &Call,
	) -> {
		self.call_index(call).map(Encode::encode)
	}
}
```

Reply:
The caller ID is provided in both the "pre-" and "post-" dispatch calls, so you won't need to pass it through.
But the "Balance" here is supposed to be the "upfront" fee we charge.

I would suggest that the upfront fee be made configurable, together with the scaling rate - since this acts also as a "ceiling" - the maximum amount fo gas fee can be changed to the user (due to how the `Imbalance` mechanism works) so this is an important parameter.

Reply:
I agree, but unfortunately the 'up-front' fee can't be made configurable if we want to avoid storage access in the pre-dispatch part...